patches/server/0726-Expose-firework-item-directly-manually-setting-fligh.patch

```diff
@@ -0,0 +1,126 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
+Date: Tue, 22 Jun 2021 23:41:11 -0400
+Subject: [PATCH] Expose firework item directly + manually setting flight ticks
+
+
+diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftFirework.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftFirework.java
+index be86114eac3975b82ca74d4d6ed3f0402a642e8a..7c8108708d40a8bd90134a6b152ef4ae9b35f1da 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftFirework.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftFirework.java
+@@ -14,24 +14,26 @@ import org.bukkit.inventory.meta.FireworkMeta;
+ public class CraftFirework extends CraftProjectile implements Firework {
+ 
+     private final Random random = new Random();
+-    private final CraftItemStack item;
++    //private final CraftItemStack item; Paper - Expose firework item directly
+ 
+     public CraftFirework(CraftServer server, FireworkRocketEntity entity) {
+         super(server, entity);
+ 
+-        ItemStack item = this.getHandle().getEntityData().get(FireworkRocketEntity.DATA_ID_FIREWORKS_ITEM);
+-
+-        if (item.isEmpty()) {
+-            item = new ItemStack(Items.FIREWORK_ROCKET);
+-            this.getHandle().getEntityData().set(FireworkRocketEntity.DATA_ID_FIREWORKS_ITEM, item);
+-        }
+-
+-        this.item = CraftItemStack.asCraftMirror(item);
+-
+-        // Ensure the item is a firework...
+-        if (this.item.getType() != Material.FIREWORK_ROCKET) {
+-            this.item.setType(Material.FIREWORK_ROCKET);
+-        }
++        // Paper Start - Expose firework item directly
++//        ItemStack item = this.getHandle().getEntityData().get(FireworkRocketEntity.DATA_ID_FIREWORKS_ITEM);
++//
++//        if (item.isEmpty()) {
++//            item = new ItemStack(Items.FIREWORK_ROCKET);
++//            this.getHandle().getEntityData().set(FireworkRocketEntity.DATA_ID_FIREWORKS_ITEM, item);
++//        }
++//
++//        this.item = CraftItemStack.asCraftMirror(item);
++//
++//        // Ensure the item is a firework...
++//        if (this.item.getType() != Material.FIREWORK_ROCKET) {
++//            this.item.setType(Material.FIREWORK_ROCKET);
++//        }
++        // Paper End - Expose firework item directly
+     }
+ 
+     @Override
+@@ -51,17 +53,28 @@ public class CraftFirework extends CraftProjectile implements Firework {
+ 
+     @Override
+     public FireworkMeta getFireworkMeta() {
+-        return (FireworkMeta) this.item.getItemMeta();
++        // Paper Start - Expose firework item directly
++        org.bukkit.inventory.ItemStack itemStack = getItem();
++        if (itemStack.getType() != Material.FIREWORK_ROCKET) {
++            return (FireworkMeta) org.bukkit.craftbukkit.inventory.CraftItemFactory.instance().getItemMeta(Material.FIREWORK_ROCKET); // Return normal meta
++        }
```

Comment:
I would get the itemstack's meta, and check its class instead of relying on the Material matching the meta. More future proof that way. Then if it is a firework meta instance, return that, else create a new meta for the firework rocket material